### PR TITLE
Parallel Resource Processing

### PR DESCRIPTION
**Pull request by damoodamoo**
# PR for issue #514 

Simple mechanism to allow the resource processor to run in multiple processes. Locally this defaults to 1, and in the build defaults to 2.

### CODE REVIEW

**Review by marrobi — COMMENTED** [review]
Are we sure this won't break anything.

For example if we try to do two firewall updates on parallel? Do we need to keep each resource serial but multiple resources parallel?

**Comment by damoodamoo** [comment]
@marrobi after this stuff could definitely break, yea - by trying to update the same resource twice in parallel. If this happened one would fail due to not getting the state lock. We don't really have a retry mechanism for an operation so the `PATCH` / `POST` would need to be resubmitted. However we have tasks as part of this story which we'll come to soon which will ensure updates to each resource happen in serial.

**Comment by marrobi** [comment]
OK, as long as its coming soon. Also one to maybe watch is if say two resources of the same type - say workspaces - that add a vnet per to core, could that cause a conflict? We need to be clear about what can and cannot happen in parallel.

File: templates/core/terraform/resource_processor/vmss_porter/cloud-config.yaml [hunk 1]
```diff
@@ -39,6 +39,7 @@ write_files:
       AZURE_TENANT_ID=${arm_tenant_id}
       ARM_USE_MSI=true
       APPLICATIONINSIGHTS_CONNECTION_STRING=${app_insights_connection_string}
+      NUMBER_PROCESSES=2
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply templates/core/terraform/resource_processor/vmss_porter/cloud-config.yaml]: done this now

File: resource_processor/shared/config.py [hunk 2]
```diff
@@ -12,6 +12,12 @@ def get_config() -> dict:
     config["resource_request_queue"] = os.environ["SERVICE_BUS_RESOURCE_REQUEST_QUEUE"]
     config["service_bus_namespace"] = os.environ["SERVICE_BUS_FULLY_QUALIFIED_NAMESPACE"]
     config["vmss_msi_id"] = os.environ.get("VMSS_MSI_ID", None)
+    config["number_processes"] = os.environ.get("NUMBER_PROCESSES", "1")
+
+    try:
+        config["number_processes_int"] = int(config["number_processes"])
+    except ValueError:
+        config["number_processes_int"] = 1
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply resource_processor/shared/config.py:21]: good spot - have done

**Review by ross-p-smith — APPROVED** [review]

File: resource_processor/vmss_porter/runner.py [hunk 3]
```diff
@@ -180,34 +181,43 @@ async def get_porter_outputs(msg_body, message_logger_adapter):
         return True, outputs_json
 
 
-async def runner():
+async def runner(i):
```
Comment (tanya-borisova) [inline resource_processor/vmss_porter/runner.py:184]: Very nit: I would rename `i` to something like `process_num`

File: templates/core/terraform/resource_processor/vmss_porter/cloud-config.yaml [hunk 4]
```diff
@@ -39,6 +39,7 @@ write_files:
       AZURE_TENANT_ID=${arm_tenant_id}
       ARM_USE_MSI=true
       APPLICATIONINSIGHTS_CONNECTION_STRING=${app_insights_connection_string}
+      NUMBER_PROCESSES=${resource_processor_number_processes_per_instance}
```
Comment (tanya-borisova) [inline templates/core/terraform/resource_processor/vmss_porter/cloud-config.yaml:42]: Very very very nit: I think it's a bit more readable when "number" goes to the end of the name, e.g. "PROCESSES_NUMBER" (similar how units are usually put at the end)

**Review by tanya-borisova — APPROVED** [review]
LGTM! Some very small naming suggestions, feel free to ignore them

**Pull request merged**